enclave/core/sgx/entropy.c

```diff
@@ -1,48 +1,86 @@
+#include <openenclave/internal/raise.h>
 #include <openenclave/internal/rdrand.h>
+#include <openenclave/internal/rdseed.h>
+#include "cpuid.h"
 
-/* TODO: This should use RDSEED instead. See issue #242. */
-oe_result_t oe_get_entropy(void* output, size_t len)
+typedef uint64_t (*_entropy_function_t)(void);
+
+OE_INLINE bool _has_cpuid_feature(
+    uint32_t leaf,
+    uint32_t feature,
+    uint32_t feature_register)
 {
-    oe_result_t ret = OE_UNEXPECTED;
-    unsigned char* p = (unsigned char*)output;
+    oe_assert(feature_register < OE_CPUID_REG_COUNT);
+    uint64_t r[OE_CPUID_REG_COUNT] = {0};
+    r[OE_CPUID_RAX] = leaf;
+    return (
+        (oe_emulate_cpuid(
+             &r[OE_CPUID_RAX],
+             &r[OE_CPUID_RBX],
+             &r[OE_CPUID_RCX],
+             &r[OE_CPUID_RDX]) == 0) &&
+        (r[feature_register] & feature));
+}
 
-    if (!output)
-        goto done;
+static oe_entropy_kind_t _get_entropy_kind()
+{
+    oe_entropy_kind_t result = OE_ENTROPY_KIND_NONE;
 
-    /* Copy 64-bit random integers to output */
-    {
-        size_t n = len / sizeof(uint64_t);
+    /* The ordering of checks is important: we want the presence of
+     * stronger entropy sources to supersede the weaker ones, so
+     * go from least to most preferred sources.
+     */
+    if (_has_cpuid_feature(1, OE_CPUID_RDRAND_FEATURE, OE_CPUID_RCX))
+        result = OE_ENTROPY_KIND_RDRAND;
 
-        while (n--)
-        {
-            uint64_t x = oe_rdrand();
+    if (_has_cpuid_feature(7, OE_CPUID_RDSEED_FEATURE, OE_CPUID_RBX))
+        result = OE_ENTROPY_KIND_RDSEED;
 
-            if (oe_memcpy_s(p, len, &x, sizeof(uint64_t)) != OE_OK)
-                goto done;
+    return result;
+}
 
-            p += sizeof(uint64_t);
-            len -= sizeof(uint64_t);
-        }
-    }
+oe_result_t oe_get_entropy(void* output, size_t len, oe_entropy_kind_t* kind)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    _entropy_function_t get_entropy = NULL;
+    unsigned char* p = (unsigned char*)output;
+    size_t bytes_left = len;
 
-    /* Copy remaining random bytes to output */
-    {
-        size_t r = len % sizeof(uint64_t);
-        uint64_t x = oe_rdrand();
-        const unsigned char* q = (const unsigned char*)&x;
+    if (kind)
+        *kind = OE_ENTROPY_KIND_NONE;
+
+    if (output)
+        memset(output, 0, len);
 
-        if (oe_memcpy_s(p, len, q, r) != OE_OK)
-            goto done;
+    if (!output || !kind)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    *kind = _get_entropy_kind();
+    if (*kind == OE_ENTROPY_KIND_RDSEED)
+        get_entropy = oe_rdseed;
+    else if (*kind == OE_ENTROPY_KIND_RDRAND)
+        get_entropy = oe_rdrand;
+    else
+        OE_RAISE(OE_UNSUPPORTED);
+
+    while (bytes_left > 0)
+    {
+        uint64_t random = get_entropy();
+        size_t copy_size =
+            (sizeof(random) > bytes_left) ? bytes_left : sizeof(random);
+        memcpy(p, &random, copy_size);
```

Comment:
`memcpy` is a special function that is aggressively optimized by compilers.
If the size is a constant, memcpy is often emitted as mov instructions.

Therefore it is better to split this into:
```c
while (bytes_left > sizeof(uint64_t)
{
    // generate and copy random uint64_t value
   // memcpy will be optimized out
    bytes_left -= sizeof(uint64_t);
}
if (bytes_left > 0)
{
    // generate last 0-7 bytes.
    // memcpy will not be optimized out
}
```

Reply:
That's disabled by -fno-builtin though: https://github.com/openenclave/openenclave/blob/de52efdfd7b15d2e04f280632f7b61d0e5d3553b/enclave/core/CMakeLists.txt#L228

https://gcc.gnu.org/onlinedocs/gcc/C-Dialect-Options.html

Reply:
> That's disabled by -fno-builtin though: 

Good point. 

One thing to do to get the full benefit of the compiler optimizations would be to call `__builtin_memcpy` here. The compiler will perform the optimizations/inlining where possible, and resort to `memcpy` where it can't.

Note that OE does not have a requirement that memcpy should not be optimized; and it is not a given that we will continue to compile with -fno-builtin.

Reply:
I would explicitly not micro-optimize for performance here. The `oe_get_entropy` codepath should not be a hot spot in the code that requires this level of hand-tuning against compiler behaviors.

Empirically, the existing code vs. the proposed modification including both the loop restructure and use of `__builtin_memcpy` compiled with Release settings:
- At 1 mil iterations of `oe_get_entropy` averaged over 100x, both result in a runtime of 0.11s.
- The variance in both runs ranges between 0.10-0.12s which suggests that any effect of the optimization is well within the bounds of noise.

Again, here my preference is for less code than more optimized code so I will keep the simplified loop structure.

@anakrish, a couple of notes:
- The reasons for optimizations as in the original code need to be called out in the comments, otherwise it's possible they will be removed in future edits. In this case, I had actually done a previous perf comparison (albeit without `__builtin_memcpy` to circumvent `-fno-builtin`, unlike in this run) and concluded that there was no reason not to simplify the loop structure, but as the group of committers grows, we should be very conscientious in documenting intention behind special constructions of code.
- The broader issue of why we compile with `-fno-builtin` is more important to address rather than specific optimizations in an instance of code. If you are passionate about this, I'd direct your attention to #1766 instead, where Maik also suggested enabling broader support for just builtins.